tests/jballerina-unit-test/src/test/resources/test-src/types/readonly/test_selectively_immutable_type.bal

```diff
@@ -147,6 +147,55 @@ function testSimpleInitializationForSelectivelyImmutableListTypes() {
     assertFalse(arr.isReadOnly());
 }
 
+function testCastOfReadonlyArrayToByteArray() {
+
+    readonly & int[] a = [1, 255];
+    any b = a;
+    byte[] c = <byte[]> b;
+    assertTrue(c === b);
+    assertEquality("[1,255]", c.toString());
+    assertEquality(1, c[0]);
+    assertEquality(255, c[1]);
+
+    float|byte[] d = <float|byte[]> b;
+    assertTrue(d === b);
+    assertEquality("[1,255]", d.toString());
+
+    (int|float|byte[])[] i = <(int|float|byte[])[]> b;
+    assertTrue(i === b);
+    assertEquality("[1,255]", i.toString());
+
+    readonly & any x = [1, 2, 3];
+    any xx = x;
+    byte[] y = <byte[]> xx;
+    assertTrue(xx === y);
+    assertEquality(1, y[0]);
+
+    readonly & int[] e = [1, 100, 1000];
+    any f = e;
+    byte[]|error g = trap <byte[]> f;
+    assertTrue(g is error);
+    error err = <error> g;
+    assertEquality("{ballerina}TypeCastError", err.message());
+    assertEquality("incompatible types: 'int[] & readonly' cannot be cast to 'byte[]'", <string> checkpanic err.detail()["message"]);
+
+
+    (float|byte)[]|error h = trap <(float|byte)[]> b;
+    assertTrue(h is error);
+    err = <error> h;
+    assertEquality("{ballerina}TypeCastError", err.message());
+    assertEquality("incompatible types: 'int[] & readonly' cannot be cast to '(float|byte)[]'", <string> checkpanic err.detail()["message"]);
+
+    readonly & string[] s = ["1", "2"];
+    any q = s;
+    byte[]|error r = trap <byte[]> q;
+    assertTrue(r is error);
+
+    err = <error> r;
+    assertEquality("{ballerina}TypeCastError", err.message());
+    assertEquality("incompatible types: 'string[] & readonly' cannot be cast to 'byte[]'", <string> checkpanic err.detail()["message"]);
+}
+
```

Comment:
shall we have test cases for union eg `(int|any)[]` `(int|byte)[]`

Reply:
test cases added